TST use global_random_seed in `sklearn/decomposition/tests/test_sparse_pca.py`

DeaMariaLeon opened the pull request:
<!--
Thanks for contributing a pull request! Please ensure you have taken a look at
the contribution guidelines: https://github.com/scikit-learn/scikit-learn/blob/main/CONTRIBUTING.md
-->

#### Reference Issues/PRs
<!--
Example: Fixes #1234. See also #3456.
Please use keywords (e.g., Fixes) to create link to the issues or pull requests
you resolved, so that they will automatically be closed when your pull request
is merged. See https://github.com/blog/1506-closing-issues-via-pull-requests
-->
Towards #22827

#### What does this implement/fix? Explain your changes.


#### Any other comments?

I wonder why `test_mini_batch_fit_transform` is kept. It is skipped all the time. It comes from PR #12253 

cc @glemaitre 

<!--
Please be aware that we are a loose team of volunteers so patience is
necessary; assistance handling other issues is very welcome. We value
all user contributions, no matter how minor they are. If we are slow to
review, either the pull request needs some benchmarking, tinkering,
convincing, etc. or more likely the reviewers are simply busy. In either
case, we ask for your understanding during the review process.
For more information, see our FAQ on this topic:
https://scikit-learn.org/dev/faq.html#why-is-my-pull-request-not-getting-any-attention.

Thanks for contributing!
-->

DeaMariaLeon committed:
Adding global random seed

DeaMariaLeon committed:
Run the CI for tests that take global_random_seed [all random seeds]
test_fit_transform
test_fit_transform_parallel
test_fit_transform_tall
test_initialization
test_scaling_fit_transform
test_pca_vs_spca
test_sparse_pca_inverse_transform
test_transform_inverse_transform_round_trip

DeaMariaLeon commented:
Testing with all the seed values were passing locally.

lucyleeow reviewed (approved):
LGTM!

(no idea about test_mini_batch_fit_transform though ...)

jeremiedbb commented on sklearn/decomposition/tests/test_sparse_pca.py:
```diff
@@ -168,32 +178,34 @@ def test_mini_batch_fit_transform():
     Z, _, _ = generate_toy_data(3, 10, (8, 8), random_state=rng)
-    spca = SparsePCA(alpha=0, ridge_alpha=0, n_components=2)
-    pca = PCA(n_components=2)
+    spca = SparsePCA(
+        alpha=0, ridge_alpha=0, n_components=2, random_state=global_random_seed
+    )
+    pca = PCA(n_components=2, random_state=global_random_seed)
```
It's not required to start from the same seedin the test, is it ?
```suggestion
    spca = SparsePCA(
        alpha=0, ridge_alpha=0, n_components=2, random_state=rng
    )
    pca = PCA(n_components=2, random_state=rng)
```

DeaMariaLeon replied:
Thank you. 
I need to ask (maybe @glemaitre ?) why sometimes it's ok to use `rng` (from np.random.RandomState(global_random_seed)`)..  and sometimes the same seed.

jeremiedbb replied:
`global_random_seed` is a number like 0, 1, 42, ...
`np.random.RandomState` is an object with a state. When you ask it to generate a random value its state changes, so the next time you ast it again it gives a different value.
If you pass an integer as random state to an estimator, it's used to create and seed a RandomState object. Therefore, each time it's fitted, it starts from the same seed and produces the same results.

So to answer your question, it depends if need to generate multiple times the exact same sequence of generated value or if it doesn't matter.
For instance, if you want to compare the results of an estimator between float32 and float64 input, you want that both estimators use the same sequence of generated values.

DeaMariaLeon replied:
Thank you very much @jeremiedbb... I see that my question wasn't clear.

In the file modified with this PR, some tests set `SparcePCA` parameter `random_state` as `random_state=rng`, where `rng = np.random.RandomState(global_random_seed)`.  For example `test_fit_transform_tall`. 

But some tests set the `random_state` parameter as `random_state=global_random_seed`. Like `test_sparse_pca_numerical_consistency`. 

I guess it doesn't matter as there is this `check_random_state(seed)` that  "Turn seed into a np.random.RandomState instance". 

In the end they might both be the same, except that changing random_state of `test_sparse_pca_numerical_consistency` to `random_state=rng` was failing. 

Anyway, your previous explanation confirmed my understanding. :-)

jeremiedbb reviewed (commented):
I think you can add ``global_random_seed`` to
- ``test_transform_nan``
- ``test_sparse_pca_numerical_consistency``

jeremiedbb commented:
> I wonder why test_mini_batch_fit_transform is kept. It is skipped all the time.

Let's just remove it. It's useless to keep dead code in the code base.

DeaMariaLeon committed:
Add global seed to more tests

DeaMariaLeon committed:
Merge remote-tracking branch 'upstream/main' into tests

DeaMariaLeon commented:
> I think you can add `global_random_seed` to
> 
> * `test_transform_nan`
> * `test_sparse_pca_numerical_consistency`

I added the seed to those. 
`test_sparse_pca_numerical_consistency` was failing a lot and I tried may things.. maybe now it's too sparse.. ?

jeremiedbb committed:
[all random seeds]
test_fit_transform
test_fit_transform_parallel
test_transform_nan
test_fit_transform_tall
test_initialization
test_scaling_fit_transform
test_pca_vs_spca
test_sparse_pca_numerical_consistency
test_sparse_pca_inverse_transform
test_transform_inverse_transform_round_trip

jeremiedbb commented:
> test_sparse_pca_numerical_consistency was failing a lot and I tried many things.. maybe now it's too sparse.. ? - the "many things" were increasing rtol, adding atol, etc.

Looking a bit, it seems to be an issue with the convergence criterion of DictionaryLearning (which is used by SparsePCA). It's not protected against rounding errors leading to negative values, so it may happen that in one case the algorithm keeps running for a few more iterations than in the other case.

It's not a major issue and not really related to this PR so I just switched the dataset for a more easy one for SparsePCA.

jeremiedbb committed:
cln [azure parallel]

jeremiedbb reviewed (approved):
LGTM. Thanks
